rhp/v4/server.go

```diff
@@ -708,61 +707,49 @@ func (s *Server) handleRPCRenewContract(stream net.Conn) error {
+			Address: renewal.NewContract.RenterOutput.Address,
+			Value:   renterInputSum.Sub(renterCost),
+		})
 	}
 
-	fceBasis, fce, err := s.contractor.ContractElement(req.Renewal.ContractID)
+	elementBasis, fce, err := s.contractor.ContractElement(req.Renewal.ContractID)
 	if err != nil {
 		return fmt.Errorf("failed to get contract element: %w", err)
 	}
-	renewalTxn.FileContractResolutions[0].Parent = fce
 
-	basis := cs.Index // start with a decent basis and overwrite it if a setup transaction is needed
-	var renewalParents []types.V2Transaction
-	if !hostCost.IsZero() {
-		// fund the locked collateral
-		setupTxn := types.V2Transaction{
-			SiacoinOutputs: []types.SiacoinOutput{
-				{Address: renewal.NewContract.HostOutput.Address, Value: hostCost},
-			},
-		}
-
-		var toSign []int
-		basis, toSign, err = s.wallet.FundV2Transaction(&setupTxn, hostCost, true)
-		if errors.Is(err, wallet.ErrNotEnoughFunds) {
-			return rhp4.ErrHostFundError
-		} else if err != nil {
-			return fmt.Errorf("failed to fund transaction: %w", err)
-		}
-		// sign the transaction inputs
-		s.wallet.SignV2Inputs(&setupTxn, toSign)
-
-		basis, renewalParents, err = s.chain.V2TransactionSet(basis, setupTxn)
-		if err != nil {
-			return fmt.Errorf("failed to get setup transaction set: %w", err)
-		}
-
-		renewalTxn.SiacoinInputs = append(renewalTxn.SiacoinInputs, types.V2SiacoinInput{
-			Parent: renewalParents[len(renewalParents)-1].EphemeralSiacoinOutput(0),
-		})
-		s.wallet.SignV2Inputs(&renewalTxn, []int{len(renewalTxn.SiacoinInputs) - 1})
+	basis, toSign, err := s.wallet.FundV2Transaction(&renewalTxn, hostCost, true)
+	if errors.Is(err, wallet.ErrNotEnoughFunds) {
+		return rhp4.ErrHostFundError
+	} else if err != nil {
+		return fmt.Errorf("failed to fund transaction: %w", err)
 	}
 
-	// update renter input basis to reflect our funding basis
+	// update renter inputs to reflect our chain basis
 	if basis != req.Basis {
-		if err := updateSiacoinElementBasis(s.chain, req.Basis, basis, renewalTxn.SiacoinInputs[:len(req.RenterInputs)]); err != nil {
+		hostInputs := renewalTxn.SiacoinInputs[len(renewalTxn.SiacoinInputs)-len(req.RenterInputs):]
+		renewalTxn.SiacoinInputs = renewalTxn.SiacoinInputs[:len(renewalTxn.SiacoinInputs)-len(req.RenterInputs)]
+		updated, err := s.chain.UpdateV2TransactionSet([]types.V2Transaction{renewalTxn}, req.Basis, basis)
+		if err != nil {
 			return errorBadRequest("failed to update renter inputs from %q to %q: %v", req.Basis, basis, err)
 		}
+		renewalTxn = updated[0]
+		renewalTxn.SiacoinInputs = append(renewalTxn.SiacoinInputs, hostInputs...)
 	}
-	// update the file contract element to reflect the funding basis
-	if fceBasis != basis {
-		if err := updateStateElementBasis(s.chain, fceBasis, basis, &fce.StateElement); err != nil {
-			return errorBadRequest("failed to update file contract basis: %v", err)
+
+	if elementBasis != basis {
+		if err := s.chain.UpdateStateElement(&fce.StateElement, elementBasis, basis); err != nil {
```

Comment:
could we hack around the need for `UpdateStateElement` this by putting the contract in a transaction and using `UpdateV2TransactionSet`?

Reply:
We could, just felt awkward. The duplicated logic is lame as well.